packages/ckeditor5-engine/src/conversion/upcasthelpers.js

```diff
@@ -934,11 +936,17 @@ function setAttributeOn( modelRange, modelAttribute, shallow, conversionApi ) {
 
 	// Set attribute on each item in range according to Schema.
 	for ( const node of Array.from( modelRange.getItems( { shallow } ) ) ) {
-		if ( conversionApi.schema.checkAttribute( node, modelAttribute.key ) ) {
-			conversionApi.writer.setAttribute( modelAttribute.key, modelAttribute.value, node );
+		if ( !conversionApi.schema.checkAttribute( node, modelAttribute.key ) ) {
+			continue;
+		}
 
-			result = true;
+		// Do not override the attribute if it's already present. See #8921.
+		if ( node.hasAttribute( modelAttribute.key ) ) {
+			continue;
 		}
```

Comment:
Maybe we should set `result = true` before this `if`, so despite that, we didn't really set element attribute, we are still consuming it's allowed by the schema, so upcast helpers ensure converter priority?

Reply:
Makes sense, but TBH, I wonder what would happen if we would consume always, not only when something was actually applied.

TBH, the converter tried to do its job so the attribute was processed. If there is another converter that could do this job better it should be added with a higher priority and overwrite the "more default" converter.

Let's say that we have such scenario:

```html
<span class="X">Foo <Y></Y> Bar</span>
```

And let's say that class `X` can be "set on text" but not on `<Y>`. Even if there is a converter that would handle setting class `X` on `<Y>`, it wouldn't work, cause text is there, so class `X` is consumed (you may say it is consumed for span, text and `Y` element).

If there's a different order of converters, and there is converter that only handles `<Y>`, then you wouldn't have conversion for text.

So, it looks to me that consuming only if something was applied is not needed cause it does not guarantee a better outcome.

Not sure about manual testing but I actually commented out the `if` here:

```javascript
// if ( attributeWasSet ) {
	conversionApi.consumable.consume( data.viewItem, match.match );
// }
```

And all tests passed.